snakefile.py

```diff
@@ -147,22 +161,35 @@ def tool(name):
             for name in generated:
                 print("  - {}".format(name))
 
+# function to pass read files to trim/filter/qc improvement
+def trim_reads_input(args):
+  sample = args[0]
+  return [os.path.join(READS_DIR, f) for f in lookup('name', sample, ['reads', 'reads2']) if f]
+
 
 rule prinseq:
-    input:
-        r1 = os.path.join(READS_DIR, "{sample}_R1.fastq"),
-        r2 = os.path.join(READS_DIR, "{sample}_R2.fastq")
+    input: trim_reads_input
     output:
-        r1 = os.path.join(TRIMMED_READS_DIR, "{sample}_trimmed_R1.fastq"),
-        r2 = os.path.join(TRIMMED_READS_DIR, "{sample}_trimmed_R2.fastq")
+        r1 = os.path.join(TRIMMED_READS_DIR, "{sample}_trimmed_R1.fastq.gz"),
+        r2 = os.path.join(TRIMMED_READS_DIR, "{sample}_trimmed_R2.fastq.gz")
     params:
         len_cutoff = int(READ_LENGTH * CUT_OFF),
         output = os.path.join(TRIMMED_READS_DIR, "{sample}"),
         tmp_r1 = os.path.join(TRIMMED_READS_DIR, "{sample}_1.fastq"),
-        tmp_r2 = os.path.join(TRIMMED_READS_DIR, "{sample}_2.fastq")
+        tmp_r2 = os.path.join(TRIMMED_READS_DIR, "{sample}_2.fastq"), 
+        read_dir = READS_DIR
     log: os.path.join(LOG_DIR, 'prinseq_{sample}.log')
-    shell: "{PRINSEQ_EXEC} -fastq {input.r1} -fastq2 {input.r2} -ns_max_n 4 -min_qual_mean 30 -trim_qual_left 30 -trim_qual_right 30 -trim_qual_window 10 -out_good {params.output} -out_bad null -min_len {params.len_cutoff} >> {log} 2>&1 && mv {params.tmp_r1} {output.r1} && mv {params.tmp_r2} {output.r2}"
-
+    run:
+        # prep input for prinseq, name without gz
+        read1 = input[0][:-3]
+        read2 = input[1][:-3]
```

Comment:
Just so I'm clear on this: this works on the file names provided by users in the sample sheet.

We have no validation of these names, so we *assume* that the files end on ".gz" and that we can just throw these trailing characters away.

Reply:
This means that a user could wreck the pipeline by giving the reads the wrong name.

Reply:
ups - yea no that shouldn't happen. The user could give uncompressed read-files....at least I made Prinseq be able to deal with both.